### PR TITLE
Remove duplicate definition of buffer_t

### PR DESCRIPTION
**Pull request by steven-johnson**
We currently have three canonical definitions of buffer_t:
— code emitted in CodeGen_C.cpp
— struct definition in buffer_t.h
— struct definition in HalideRuntime.h

The first can’t be easily eliminated (yet), but the second should be
eliminated in favor of simply including the definition from
HalideRuntime.h

### CODE REVIEW

File: src/runtime/HalideRuntime.h [hunk 1]
```diff
@@ -294,7 +294,15 @@ extern void halide_memoization_cache_cleanup();
  * The raw representation of an image passed around by generated
  * Halide code. It includes some stuff to track whether the image is
  * not actually in main memory, but instead on a device (like a
- * GPU). */
+ * GPU).
```
Comment (abadams) [inline src/runtime/HalideRuntime.h]: This comment was cut intentionally because it didn't seem like it belonged in a user-facing file.

**Comment by abadams** [comment]
The rest of the stuff in HalideRuntime.h is not necessarily safe to include by users of buffer_t.h - those other functions don't exist.

I guess you could guard the rest of it with #ifndef COMPILING_HALIDE

**Comment by steven-johnson** [comment]
There are already several other places we're including HalideRuntime.h: Buffer.cpp, JITModule.h, Tracing.cpp... I'm happy to add the #ifndef but this statement gives me pause.

**Comment by abadams** [comment]
Hrm, good point. I guess we're already assuming it's safe.

**Comment by steven-johnson** [comment]
Let me add some ifdef'y goodness, just in case

**Comment by steven-johnson** [comment]
Fun fact: Buffer.cpp references halide_device_free, halide_copy_to_device, and halide_copy_to_host.

**Comment by steven-johnson** [comment]
Ugh, ugly can of worms; in the land of JIT, runtime vs not-runtime is a more fuzzy concept, and class Buffer blurs those boundaries quite effectively. 

IMHO adding some sanity to this file (e.g. ifdefs) is a good idea, but one that perhaps should be considered separately from this issue.

**Comment by abadams** [comment]
I just bit the bullet and included HalideRuntime.h instead of buffer_t everywhere.

**Pull request closed without merging**